runtime/gc_base/StringTable.cpp

```diff
@@ -976,4 +975,58 @@ j9gc_internString(J9VMThread *vmThread, j9object_t sourceString)
 	return internedString;
 }
 
+typedef struct {
+	J9UTF8* utf;
+	j9object_t stringObject;
+} J9UTFCacheEntry;
+
+static UDATA
+utfCacheHashFn(void *key, void *userData)
+{
+	/* UTF8 pointers are 2-aligned, so shift the pointer for maximum entropy */
+	return ((UDATA)((J9UTFCacheEntry*)key)->utf) >> 1;
+}
+
+static UDATA
+utfCacheHashEqualFn(void *leftKey, void *rightKey, void *userData)
+{
+	return ((J9UTFCacheEntry*)leftKey)->utf == ((J9UTFCacheEntry*)rightKey)->utf;
+}
+
+j9object_t
+j9gc_createJavaLangStringWithUTFCache(J9VMThread *vmThread, J9UTF8 *utf)
+{
+	j9object_t result = NULL;
+	/* Allocation may cause a GC which frees the cache - ensure no local copy of the cache
+	 * is held across the allocation.
+	 */
+	{
+		J9HashTable *utfCache = vmThread->utfCache;
+		if (NULL != utfCache) {
+			J9UTFCacheEntry exemplar = { utf, NULL };
+			J9UTFCacheEntry *entry = (J9UTFCacheEntry*)hashTableFind(utfCache, &exemplar);
+			if (NULL != entry) {
+				result = J9WEAKROOT_OBJECT_LOAD(vmThread, &entry->stringObject);
+				goto done;
+			}
+		}
+	}
+	result = j9gc_createJavaLangString(vmThread, J9UTF8_DATA(utf), J9UTF8_LENGTH(utf), J9_STR_INTERN);
+	if (NULL != result) {
+		J9HashTable *utfCache = vmThread->utfCache;
```

Comment:
Does `vmThread->utfCache` have to be declared volatile to ensure this works?  It is accessed by multiple threads...

Reply:
The only MT access to the field involves VM access, so there are already sufficient barriers there.

Reply:
Marking the field volatile be a good idea, though I don't think any compiler would actually optimize the two fetches into one across a function call (particularly one that takes a pointer to the structure as a parameter).

Reply:
Sorry, may I ask more about this? How exactly it is going to work?
There are two situations potentially for this non-atomic logic:
- race between two threads to create hash table, so what prevents two hash tables to be created?
- hypothetical race between potential creation of hash table and cleaning it by GC. Yes, it probably means GC code (particularly in Metronome where cleaning is not logically bounded) should not call this API

Reply:
ok, second case is probably impossible even for Metronome and we never execute this code concurrently

Reply:
There is one cache per thread, so there are no races. The cache is only ever used with VM access and  is cleared in the GC under exclusive access.

Reply:
Right, thank you. I missed this cache is bound to the thread for a moment. So if GC does not execute cleaning concurrently it should work.